Recipe and Input class definitions with e2e export

tarun292 opened the pull request:
Based on the discussion in https://github.com/pytorch/executorch/discussions/9027

This PR adds the executorch.export API and all the supporting components required for it. At a high level the executorch.export API takes in a model, example inputs and a recipe, then underneath the hood executes all the steps required to export, quantize and lower the model based on the recipe.

The pipeline consists of a staged setup where each major step in the process such as Export, Quantization etc. is listed as a separate stage and a chain of these is formed and then executed. The result of this will be that we'll get a PTE file which can then be executed on device.

The major new components added in this PR are:
- class definition for ExportSession and ExportRecipe
- Definitions for each stage in the process
- executorch.export API which will return a session object that the user can then use to get access to the PTE file, run the model via pybindings, print delegation info etc.

tarun292 committed:
Recipe and Input class definitions with e2e export



Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

digantdesai commented on export/_export.py:
```diff
@@ -0,0 +1,357 @@
+    return manager
+
+
+class ExportSession:
```
This is a good idea, I like it. But just a pet peeve, let's refactor this before starting, and organize the logic so that it can scale better, else I fear it will end up becoming a giant bowl of spaghetti.

Suggestion, make a `class Stage(ABC)` and derive others from it like `Quantize(Stage)` and `Export(Stage)` and then Session just manages the pipeline with `Stage`s. 

See - https://github.com/pytorch/executorch/blob/409447d75a1524c1acc8f8ea894c2e13dd723a79/backends/xnnpack/test/tester/tester.py#L424

digantdesai commented on export/_recipe.py:
```diff
@@ -0,0 +1,83 @@
+
+    name: Optional[str] = None
+    quantizer: Optional[Quantizer] = None
+    edge_compile_config: Optional[EdgeCompileConfig] = ( # pyre-ignore[11]: Type not defined
```
mypy?

digantdesai commented on export/_recipe.py:
```diff
@@ -0,0 +1,83 @@
+    )
+    mode: Mode = Mode.RELEASE
+
+    def get_quantizer(self) -> Optional[Quantizer]:
```
nit property?

digantdesai commented on export/_export.py:
```diff
@@ -0,0 +1,357 @@
+from ._recipe import ExportRecipe
+
+
+def export(
```
nit, WDYT about? I understand we want to do `s/torch.export/executorch.export` but it feels a bit confusing given we are returning something not analogous to the model which torch.export does.
```suggestion
def create_export_session(
# or
Session.__init__
# or
@classmethod
Session.create()
```

digantdesai commented on export/_export.py:
```diff
@@ -0,0 +1,357 @@
+    model: Union[nn.Module, Dict[str, nn.Module]],
+    example_inputs: Union[List[tuple[torch.Tensor, ...]], Dict[str, List[tuple[torch.Tensor, ...]]]],
+    export_recipe: ExportRecipe,
+    name: Optional[str] = None,
```
what is the rationale for this?

digantdesai commented on export/_export.py:
```diff
@@ -0,0 +1,357 @@
+        dynamic_shapes: Optional dynamic shape specifications
+        constant_methods: Optional dictionary of constant methods
+        artifact_dir: Optional directory to store artifacts
+        apply_quantization: Whether to apply quantization during export, defaults to False
```
this feels a bit confusing

digantdesai commented on export/_export.py:
```diff
@@ -0,0 +1,357 @@
+    Returns:
+        A configured ExportSession instance with the export process completed if requested
+    """
+    manager = ExportSession(
```
nit
```suggestion
    session = ExportSession(
```

digantdesai commented on export/_export.py:
```diff
@@ -0,0 +1,357 @@
+    example_inputs: Union[List[tuple[torch.Tensor, ...]], Dict[str, List[tuple[torch.Tensor, ...]]]],
+    export_recipe: ExportRecipe,
+    name: Optional[str] = None,
+    dynamic_shapes: Optional[Union[Any, Dict[str, Any]]] = None,
```
why is this not on `Session.export()`?

digantdesai reviewed (approved):
Looks good at a high level. Left some comments.

mergennachin commented on export/_export.py:
```diff
@@ -0,0 +1,357 @@
+
+from ._recipe import ExportRecipe
+
+
```
We should annotate with executorch.exir._warnings.experimental decorators

tarun292 committed:
Update base for Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

tarun292 committed:
Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

jackzhxng commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+from .recipe import ExportRecipe
+
+
+class Stage(ABC):
```
I like this 👍🏻

jackzhxng commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        pass
+
+    @abstractmethod
+    def get_outputs(self) -> Any:
```
```suggestion
    def get_artifacts(self) -> Any:
```

Think this conveys the intent of this function better when the user hasn't read the docstring

tarun292 replied:
Updated.

jackzhxng commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+
+    @property
+    def name(self) -> str:
+        return "quantize"
```
Would be worth to comment somewhere that quantization may also happen in `SourceTransformStage`

jackzhxng commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+            )
+        return self._executorch_program_manager.buffer
+
+    def get_example_input(
```
I feel like this and `run_method` seem a bit confusing to have in the ExportSession?

tarun292 replied:
The way i'm thinking is that export session lets the users do almost all the things they'd like to with an executorch model like load the model and run via pybindings, get reference inputs etc. and that's why I added these helper utils.

jackzhxng commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        elif stage_name == "executorch":
+            self._executorch_program_manager = result
+
+    def _run_pipeline(self, start_stage: str) -> None:
```
Why do we need this start_stage stuff, can we just not take any parameters and just run the all stages in self._stages?

kimishpatel replied:
Yeah I agree. We should not have this start_stage thing. We should just construct pipeline that can just be run with each stage in sequence. Construction of the pipeline should take care of apprpriate deps

tarun292 replied:
Addressed in latest version.

jackzhxng commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        return self._executorch_program_manager
+
+
+class SourceTransformStage(Stage):
```
NIT: have this and QuantizeStage declared in the order of the export process, this one declared at the top

kimishpatel commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+    ) -> None:
+        self._exported_program: Dict[str, ExportedProgram] = {}
+        self._pre_edge_transform_passes = pre_edge_transform_passes
+        self._model_dict: Dict[str, nn.Module] = {}
```
is this to capture more than one Module to be exported? why do we need this?

tarun292 replied:
Yep a user could possibly pass in multiple methods to be exported and wanted to have support for that from the beginning. Something we struggled with in ModAI by trying to add it later.

kimishpatel commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+            self._dynamic_shapes_dict = export_config.get("dynamic_shapes", {})
+
+        # Check if we need to do export (if _exported_program is empty)
+        if not self._exported_program:
```
What if you want to rexport? For different input shapes or export args? If ExportStange is strictly for one export session and to export with different args you need a new ExportStage than I will make it immutable.

tarun292 replied:
This wasn't needed, removed it. One export session is intended to be only for one set of input args.

kimishpatel commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+
+        # If quantization is available, add it after source transform
+        if "quantize" in self._stages:
+            self._pipeline.insert(0, "quantize")
```
As I said elsewhere I think quantize should not be responsible for running export. We should run export, run quantizer, prepare, convert, and run export again. This call can be managed within here by detecting the need to run export again if we have pt2e quantize requested

tarun292 replied:
We need to store this because it's used in all the later stages such as to_edge etc.

kimishpatel commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        self._current_stage_index = start_index
+
+        # Process each stage in the pipeline
+        for i in range(start_index, len(self._pipeline)):
+            stage_name = self._pipeline[i]
+            self._current_stage_index = i
+
+            # Get the primary input and configuration parameters for this stage
+            primary_input, config_params = self._get_stage_inputs(stage_name)
+
+            # Run the stage
+            result = self._run_stage(stage_name, primary_input, config_params)
+
+            # Store the result
+            self._store_stage_result(stage_name, result)
```
Yeah at thsi point i expect jsut iterate through stages and run without needing to figure out input and storage stage outputs. Outputs of each stage should be there in `get_output/artifact` if needed

tarun292 replied:
This means that the quantization recipe doesn't necessarily have configs to use with quantize_ but might still have pt2e quantization. In that case it doesn't make sense to throw an error right?

kimishpatel replied:
Not sure I follow what you mean.

kimishpatel commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        # The original code expects this to be a tuple of tensors
+        return self._example_inputs[method_name][0]
+
+    def run_method(
```
why is this part of this class

tarun292 replied:
I think that's a valid point, but i've felt like having all the quantization configuration related items in one class would generally be easier for recipe authors. We when creating the pipeline decide what source transforms need to be run and what is in pt2e. This isn't set in stone though, we'll definitely iterate on this as we add more recipes in XNNPack, CoreML etc. and we can arrive at a different structure over the next few weeks if the current one isn't flexible enough.

kimishpatel reviewed (commented):
Left quite a few comments

kimishpatel reviewed (commented):
I didnt "request changes" if you feel strongly about landing, but I think we should consider addressing these comments

jackzhxng commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+    def run(
+        self,
+        models: Dict[str, Any],
+        export_config: Optional[Dict[str, Any]] = None,
```
Looking at this PR again, why isn't this a part of the ExportRecipe?

tarun292 replied:
This contains example inputs and dynamic shapes which shouldn't be a part of the recipe.

tarun292 committed:
Update base for Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

tarun292 committed:
Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

tarun292 commented on export/__init__.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Thought quite a bit about this, the main reasoning to call this export too is because we're consuming the export call within the API. One of the main points of this API is that if someone wants to target ExecuTorch they don't even need to learn about torch.export (although many of them probably will).

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        transform_passes: Optional[Sequence[Callable[[Any], Optional[Any]]]] = None,
+        compile_config: Optional[Any] = None,
+    ) -> None:
+        self._partitioners = partitioners
+        self._transform_passes = transform_passes
```
Transform passes are what we run after to_edge e.g. edge_manager.transform(), and partitioners are the ones we use in to_backend.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        self._compile_config = compile_config
+        self._edge_program_manager: Optional[EdgeProgramManager] = None
+        self._delegation_info = None
+        self._exported_program: Dict[str, ExportedProgram] = {}
```
This is needed in later stages right after we export and do quantization and to_edge etc.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        return self._executorch_program_manager
+
+
+class SourceTransformStage(Stage):
```
Thought a bit about this, it doesn't necessarily mean all the needed source transforms will be squeezed in here, but this will be needed for quantize_ and maybe other things in the future so it made sense to have a source transform stage.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        self._transformed_models = models
+
+        # Check if there's a quantization recipe with ao_base_config
+        if self._quantization_recipe and self._quantization_recipe.ao_base_config:
```
This is an optional config right, if this is not specified we can just continue without quantizing.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        return self._executorch_program_manager
+
+
+class SourceTransformStage(Stage):
```
Thought a bit about this, structuring it this way doesn't necessarily bring us a lot of benefits as `EagerQuantRecipe` and `ExportQuantRecipe` don't necessarily have a lot of common things which they can inherit from by having a base `QuantizationRecipe` class.

kimishpatel replied:
I am against having SourceTranformStage just for quantization. It doesnt feel right. In fact quantize_ is not a source transformation. So I am not sure why quantization is wrapped up here at all.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+
+        if calibration_config is not None:
+            self._example_inputs_dict = calibration_config.get("example_inputs", {})
+            dynamic_shapes_dict = calibration_config.get("dynamic_shapes", {})
```
Yep definitely in the plans for next udpate.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+
+            self._exported_models_dict[method_name] = export_for_training(
+                model, self._example_inputs_dict[method_name][0], **kwargs
+            ).module()
```
Updated in latest version.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+            prepared_model = prepare_pt2e(captured_model, self._quantizer)  # type: ignore
+
+            # Allow the exported model to switch between train and eval modes
+            allow_exported_model_train_eval(prepared_model)
```
This is something we noticed that was needed internally for dropout and batchnorm.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+@experimental(
+    "This API and all of its related functionality such as ExportSession and ExportRecipe are experimental."
+)
+class ExportSession:
```
This is an active session where users can do things like run the model via pybindings, print delegation info etc. that's why i decided to call it session rather than pipeline.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+
+        # Create the source transform stage if a quantization recipe is provided
+        if self._export_recipe.quantization_recipe is not None:
+            self._stages["source_transform"] = SourceTransformStage(
```
Within this stage in the future i wanted to be able to support running source transforms too if needed.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+                    self._stages["quantize"] = QuantizeStage(quantizer=quantizer)
+
+        # Create the export stage
+        self._stages["export"] = ExportStage(
```
Updated in latest version.

tarun292 commented on export/export.py:
```diff
@@ -0,0 +1,866 @@
+        # Return the outputs
+        return self._stages[stage_name].get_outputs()
+
+    def _get_stage_inputs(self, stage_name: str) -> tuple[Any, Dict[str, Any]]:
```
I thought a lot about this too, in order to deal with quantize_ and later on other things such as custom op replacement in eager mode we do need to do it and it's quite a bit of complexity that'll be better hidden from the users.

kimishpatel replied:
> deal with quantize_ and later on other things such as custom op replacement in eager mode

Again eager mode should not belong to this. It is coupling unrelated things together.

tarun292 commented:
@kimishpatel @jackzhxng thanks for the detailed review, really appreciate it! Addressed all of your comments:

In the latest version the major changes i made are:
- Removed start_stage API and string input for stage name, rather just have the run_pipeline function now which runs through the entire pipeline. Changed quite a bit how the pipeline is constructed now.
- The pipeline is a list of stages now and not a dictionary.
- Export happens independently of quantization now.
- Added get_artifact which gets the output of each stage.

digantdesai commented:
Thanks @tarun292 - love the pipeline design.

tarun292 committed:
Update base for Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

tarun292 committed:
Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

tarun292 committed:
Update base for Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

tarun292 committed:
Update on "Recipe and Input class definitions with e2e export"


Differential Revision: [D71946730](https://our.internmc.facebook.com/intern/diff/D71946730/)

[ghstack-poisoned]

tarun292 committed:
Merge branch 'main' into gh/tarun292/4/head
